### PR TITLE
Suppress unused_mut if unused_variables is reported

### LINKED ISSUE #97074: Inhibit unused_mut variables if unused_variables is also triggered
Given the following code: 

```rust
fn main() {
    let mut s = String::from("a");
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0f9622ef596ee0a54b44e64f7e1579b0).)

The current output is:

```
warning: unused variable: `s`
 [--> src/main.rs:2:13
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)  |
2 |     let mut s = String::from("a");
  |             ^ help: if this is intentional, prefix it with an underscore: `_s`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: variable does not need to be mutable
 [--> src/main.rs:2:9
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)  |
2 |     let mut s = String::from("a");
  |         ----^
  |         |
  |         help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` on by default

warning: `playground` (bin "playground") generated 2 warnings
```

```
warning: unused variable: `s`
 [--> src/main.rs:2:13
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)  |
2 |     let mut s = String::from("a");
  |             ^ help: if this is intentional, prefix it with an underscore: `_s`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: variable does not need to be mutable
```

It's obvious that an unused variable does not need to be declared mutable, and the extra warning puts more cognitive burden on the reader to figure out what's relevant.

Found this on 1.60.0, but also still happens on nightly.

### PR DESCRIPTION
**Pull request by sanxiyn**
Fix #97074.

### CODE REVIEW

File: compiler/rustc_middle/src/query/mod.rs [hunk 1]
```diff
@@ -830,7 +830,8 @@ rustc_queries! {
         desc { |tcx| "checking privacy in {}", describe_as_module(key, tcx) }
     }
 
-    query check_liveness(key: DefId) {
+    query check_liveness(key: DefId) -> Option<FxIndexSet<Span>> {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/query/mod.rs:833]: Spans are not a very robust source of information. Do we have a better, more "semantic" source of information?

Comment (sanxiyn) [reply compiler/rustc_middle/src/query/mod.rs:833]: I tried but couldn't find one. Rather, there is a lot on liveness side working on HIR, but they are missing on borrowck side working on MIR, which needs to use this information.

File: compiler/rustc_passes/src/liveness.rs [hunk 2]
```diff
@@ -1594,24 +1601,29 @@ impl<'tcx> Liveness<'_, 'tcx> {
                 if ln == self.exit_ln { false } else { self.assigned_on_exit(ln, var) };
 
             if is_assigned {
+                let spans = hir_ids_and_spans
+                    .into_iter()
+                    .map(|(_, _, ident_span)| ident_span)
+                    .collect::<Vec<_>>();
+                self.unused_variables_spans.borrow_mut().extend(&spans);
```
[the remark this answers is not in the document]
Comment (sanxiyn) [reply compiler/rustc_passes/src/liveness.rs:1608]: Right. It is troublesome indeed. I will think about it.

**Comment by nagisa** [comment]
@cjgillot you want to keep reviewing this to the end or should I take a look at this still?

**Comment by cjgillot** [comment]
I can take it you don't want to keep it.
r? @cjgillot

**Comment by JohnCSimon** [comment]
@sanxiyn Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**